app/components/file_attachment_list/file_attachment_document.js

```diff
@@ -0,0 +1,292 @@
+            if (!isDir) {
+                try {
+                    await RNFetchBlob.fs.mkdir(DOCUMENTS_PATH);
+                } catch (error) {
+                    // Do nothing
+                }
+            }
+
+            const options = {
+                session: file.id,
+                timeout: 10000,
+                indicator: true,
+                overwrite: true,
+                path
+            };
+
+            const exist = await RNFetchBlob.fs.exists(path);
+            if (exist) {
+                this.openDocument(file, 0);
+            } else {
+                this.setState({downloading: true});
+                this.downloadTask = RNFetchBlob.config(options).fetch('GET', getFileUrl(file.id)).
+                    progress((received, total) => {
+                        const progress = (received / total) * 100;
+                        if (this.mounted) {
+                            this.setState({progress});
+                        }
+                    });
+
+                await this.downloadTask;
+                if (this.mounted) {
+                    this.setState({
+                        progress: 100
+                    }, () => {
+                        // need to wait a bit for the progress circle UI to update to the give progress
+                        this.openDocument(file);
+                    });
+                }
+            }
+        } catch (error) {
+            RNFetchBlob.fs.unlink(path);
+            if (this.mounted) {
+                this.setState({downloading: false, progress: 0});
+
+                if (error.message !== 'cancelled') {
+                    const {intl} = this.context;
+                    Alert.alert(
+                        intl.formatMessage({
+                            id: 'mobile.downloader.failed_title',
+                            defaultMessage: 'Download failed'
+                        }),
+                        intl.formatMessage({
+                            id: 'mobile.downloader.failed_description',
+                            defaultMessage: 'An error occurred while downloading the file. Please check your internet connection and try again.\n'
+                        }),
+                        [{
+                            text: intl.formatMessage({
+                                id: 'mobile.server_upgrade.button',
+                                defaultMessage: 'OK'
+                            })
+                        }]
+                    );
+                }
+            }
+        }
+    };
+
+    handlePreviewPress = async () => {
+        const {file} = this.props;
+        const {downloading, progress} = this.state;
+
+        if (downloading && progress < 100) {
+            this.cancelDownload();
+        } else if (downloading) {
+            this.resetViewState();
+        } else {
+            this.downloadAndPreviewFile(file);
+        }
+    };
+
+    openDocument = (file, delay = 2000) => {
```

Comment:
Why do we need a delay here, and does it need to be such a long delay? Also, I think it would be much clearer to either not use a default here or have a default of 0 since that's what I'd expect a number to default to

Reply:
no, the default here is 2 seconds cause that's the amount of time it takes to the progress bar to complete the circle, I can add a comment saying that just to be sure

Reply:
That seems like a long time to me, but alright, we can see how it feels once it's in